src/ndarray/ndarray.cc

```diff
@@ -558,6 +558,101 @@ void CopyFromTo(const NDArray& from, const NDArray& to, int priority) {
   }
 }
 
+void Quantize(const NDArray &from, NDArray *to, NDArray *residual, const std::string& compress,
+              const float neg_threshold, const float pos_threshold,
+              int priority) {
+  CHECK(from.shape().ndim() != 0)
+      << "source operands have zero dimension shape";
+  // important: callback must always capture by value
+  NDArray ret = *to;
+  NDArray res = *residual;
+  int a = from.ctx().dev_mask();
+  int b = to->ctx().dev_mask();
+  if (a == cpu::kDevMask && b == cpu::kDevMask) {
+    if (compress == "2bit") {
+      Engine::Get()->PushSync([from, res, ret, neg_threshold, pos_threshold](RunContext ctx) {
+          std::vector<TBlob> inputs(3);
```

Comment:
It will be more efficient to initialize as { from.data(), res.data(), ret.data() }, since this won't cause TBlob() constructor to be called three times when the initial size of the vector is set.

Reply:
Ok, cool. I didn't know about this